tests/models/test_admin.py

```diff
@@ -1,156 +1,180 @@
+
+def test_password_hashed(db_session, create_admin_user):
+    # GIVEN: An Admin user
+    # WHEN:  The Admin password is set to "password"
+    # THEN:  Check the hashed password starts with $2a$
+    admin = create_admin_user(db_session)
+    admin.password = "password"
+
+    pytest.raises(NotImplementedError, lambda: admin.password)
+    assert admin.password_hashed.startswith('$2a$')
+
+def test_with_password(db_session, create_admin_user):
+    # GIVEN: An Admin user
+    # WHEN:  There is no password set
+    # THEN:  The Admin.with_password query is empty
+    assert [] == Admin.with_password(db_session).all()
+
+    admin = create_admin_user(db_session)
+    assert [] == Admin.with_password(db_session).all()
+
+    # WHEN: A password is set
+    # THEN: The Admin.with_password query contains users
+    admin.password = "password"
+    assert [admin] == Admin.with_password(db_session).all()
+
+    admin2 = create_admin_user(db_session, email="admin2@nypl.org")
+    assert [admin] == Admin.with_password(db_session).all()
+
+    admin2.password = "password2"
+    assert set([admin, admin2]) == set(Admin.with_password(db_session).all())
+
+def test_with_email_spaces(db_session, create_admin_user):
+    # GIVEN: An Admin user
+    # WHEN:  Creating the Admin user, the email address contains spaces
+    # THEN:  The email address should not contain spaces
+    admin_spaces = create_admin_user(db_session, email="test@email.com ")
+    assert "test@email.com" == admin_spaces.email
+
+def test_has_password(db_session, create_admin_user):
+    # GIVEN: An Admin user
+    # WHEN:  Setting the password to "password"
+    # THEN:  The password should be "password" and not "banana"
+    admin = create_admin_user(db_session)
+    admin.password = "password"
+    assert True == admin.has_password("password")
+    assert False == admin.has_password("banana")
+
+def test_authenticate(db_session, create_admin_user):
+    # GIVEN: Two Admin users with different passwords
+    # WHEN:  Authenticating the Admin users with the password "password"
+    # THEN:  The Admin user that has the password "password" is authenticated
+    admin = create_admin_user(db_session, email="admin@nypl.org")
+    admin.password = "password"
+    other_admin = create_admin_user(db_session, email="other@nypl.org")
+    other_admin.password = "banana"
+
+    assert admin == Admin.authenticate(db_session, "admin@nypl.org", "password")
+    assert None == Admin.authenticate(db_session, "other@nypl.org", "password")
+    assert None == Admin.authenticate(db_session, "example@nypl.org", "password")
+
+def test_roles(db_session, create_admin_user, create_library):
+    # GIVEN: An Admin user and a Library
+    # WHEN:  Changing the Admin user roles
+    # THEN:  The Admin user should have appropriate access based on role
+    admin = create_admin_user(db_session)
+    library = create_library(db_session)
+    assert False == admin.is_system_admin()
```

Comment:
If you're testing for a boolean output, it's a good idea to use the `is` comparator rather than `==`. In the vast majority of cases it won't matter, but it's worth noting that there's some hidden type coercion that happens when you use the double equals with a boolean value. Since the double equals isn't a boolean operator, all boolean values passed to it end up converted to the integers 1 and 0 before comparison happens. So you can get this situation (that's oddly reminiscent of JavaScript's double vs. triple equals):

```Python
❯ python -W ignore
>>> True is True
True
>>> True is 1
False
>>> True == 1
True
```

There's two valid alternatives to `assert myobj.mymethod() == False`:

1. `assert myobj.mymethod() is False`
2. `assert not myobj.mymethod()`

I think most Python sources favor the second method any time you're testing the truth of a single value, and would recommend only using the `is` comparator when you have to determine whether two unknown booleans are the same. That said, I actually prefer the first approach in a test context, because it very clearly states the expected outcome, and doesn't require extra thought on the part of the reader.

Also (and this is largely personal style, but a thing I Have Opinions About), I think it's more readable to have assertions structured left to right as:

```Python
assert actual_value == expected_value     # or whatever comparator is appropriate, if not ==
```

rather than `expected` before `actual`. I don't think there's a definitive rule about this anywhere, so feel free to disagree if you also Have Opinions. FWIW, Guido van Rossum [said in 2010](https://mail.python.org/pipermail/python-dev/2010-December/106954.html) that thinking about it in terms of "expected" and "actual" isn't good, and we should just use "first" and "second". But whatever, that dude's wrong.